### PR TITLE
CAKE03 - Large LP Deposit/Withdrawal

### PR DESCRIPTION
**Pull request by sprtd**
#  Large Liquidity Pool Deposit/Withdrawal Bot

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [ ] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this bot please describe it here.

### CODE REVIEW

File: PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    - `poolAddress`: The address of the liquidity pool 
+    - `token0`: The address of token0
+    - `token1`: The address of token1
+    - `amount0`: The withdrawn amount of token0
+    - `amount1`: The withdrawn amount of token1
+    - `totalSupply`: The total supply of the pool before the deposit transaction at the preceeding block
+    
+## Test Data
+
+The agent behaviour can be verified with the following transactions:
```
Comment (Vxatz) [inline PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/README.md:43]:
```suggestion
The bot behaviour can be verified with the following transactions:
```

File: PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/package.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    "publish": "forta-agent publish",
+    "push": "forta-agent push",
+    "disable": "forta-agent disable",
+    "enable": "forta-agent enable",
+    "keyfile": "forta-agent keyfile",
+    "test": "jest --detectOpenHandles",
+    "format": "prettier --write \"src/**/*.ts\""
+  },
+  "dependencies": {
+    "forta-agent": "^0.1.7",
```
Comment (Vxatz) [inline PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/package.json:27]:
```suggestion
    "forta-agent": "^0.1.8",
```

Please use the updated version as it doesn't have the chanId issues

File: PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    const block: number = txEvent.blockNumber;
+    const logs: LogDescription[] = txEvent.filterLog(EVENTS_ABI);
+    if (!logs) return findings;
+
+    await Promise.all(
+      logs.map(async (log) => {
+        const { amount0, amount1 } = log.args;
+        const [valid, tokenA, tokenB, totalSupply] = await fetcher.getPoolData(block - 1, log.address);
+        let token0: string = tokenA < tokenB ? tokenA : tokenB;
+        let token1: string = tokenA < tokenB ? tokenB : tokenA;
```
Comment (Vxatz) [inline PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/src/agent.ts:27]: Actually, you don't need these as the fetcher already returns token0, token1, not tokenA, tokenB

File: PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/src/agent.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        const createdPair = createPair(token0, token1, factory);
+        if (valid && log.address === createdPair) {
+          const [balance0, balance1] = await fetcher.getPoolBalance(block - 1, log.address, token0, token1);
+          if (
+            totalSupply.gt(poolSupplyThreshold) &&
```
Comment (Vxatz) [inline PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/src/agent.ts:32]:
```suggestion
            totalSupply.gt(poolSupplyThreshold) &&
```

This is a nice addition, but it should be declared more clearly in the README both in the Description and in the Alerts section

+This requires an extra test

File: PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/src/finding.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    token1: token1,
+    amount0: log.args.amount0.toString(),
+    amount1: log.args.amount1.toString(),
+    totalSupply: totalSupply.toString(),
+  };
+
+  if (log.name === "Mint") {
+    return Finding.fromObject({
+      name: "Large LP Deposit in Pancakeswap pool",
+      description: `${log.name} event with large amounts emitted from Pancakeswap pool`,
```
Comment (Vxatz) [inline PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/src/finding.ts:17]:
```suggestion
      description: `${log.name} event with large amounts emitted from a Pancakeswap pool`,
```

File: PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/src/finding.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      alertId: "CAKE-3-1",
+      severity: FindingSeverity.Info,
+      type: FindingType.Info,
+      protocol: "Pancakeswap",
+      metadata,
+    });
+  } else
+    return Finding.fromObject({
+      name: "Large LP Withdrawal from Pancakeswap pool",
+      description: `${log.name} event with large amount emitted from an Pancakeswap pool`,
```
Comment (Vxatz) [inline PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/src/finding.ts:27]:
```suggestion
      description: `${log.name} event with large amounts emitted from a Pancakeswap pool`,
```

Just keeping consistency between the two descriptions.

**Review by Vxatz — CHANGES_REQUESTED** [review]

File: PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/README.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+## Supported Chains
+
+- Binance Smart Chain 
+
+## Alerts
+
+- CAKE-3-1
+
+  - Fired when large liquidity is deposited to a Pancakeswap pool as `Mint` event is emitted with a `totalSuppply` that exceeds the pool's `poolSupplyThreshold` and `amount0`/`amount1` that exceeds `token0`/`token1` `thresholdPercentage`
```
Comment (Vxatz) [inline PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/README.md:18]:
```suggestion
- CAKE-3-1
  - Fired when large liquidity is deposited to a Pancakeswap pool as `Mint` event is emitted with a `totalSuppply` that exceeds the pool's `poolSupplyThreshold` and `amount0`/`amount1` that exceeds `token0`/`token1` `thresholdPercentage`
```

(just removing the space)

File: PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/README.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    - `amount0`: The deposited amount of token0
+    - `token1`: The address of token1
+    - `amount1`: The deposited amount of token1
+    - `totalSupply`: The total supply of the pool before the deposit transaction at the preceeding block
+    
+
+- CAKE-3-1
+  - Fired when large liquidity is withdrawn from a Pancakeswap pool as `Mint` event is emitted with a `totalSuppply` that exceeds the pool's `poolSupplyThreshold` threshold and `amount0`/`amount1` that exceeds `token0`/`token1` `thresholdPercentage`
+  - Severity is always set to "Medium"
+  - Type is always set to "Suspicious"
```
Comment (Vxatz) [inline PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/README.md:33]:
```suggestion
- CAKE-3-2
  - Fired when large liquidity is withdrawn from a Pancakeswap pool as `Burn` event is emitted with a `totalSuppply` that exceeds the pool's `poolSupplyThreshold` threshold and `amount0`/`amount1` that exceeds `token0`/`token1` `thresholdPercentage`
  - Severity is always set to "Info"
  - Type is always set to "Info"
```

File: PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/package.json [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    "format": "prettier --write \"src/**/*.ts\""
+  },
+  "dependencies": {
+    "forta-agent": "^0.1.8",
+    "lru-cache": "^7.10.2"
+  },
+  "devDependencies": {
+    "@types/jest": "^27.0.1",
+    "@types/nodemon": "^1.19.0",
+    "forta-agent-tools": "^2.2.1",
```
Comment (Vxatz) [inline PancakeSwap-Bots/Large-LP-Deposit-Withdrawal/package.json:33]:
```suggestion
    "forta-agent-tools": "^2.2.2",
```

Please 1) delete package-lock.json 2) change this to `^2.2.2` 3) run `npm install` and 4) push

**Review by Vxatz — APPROVED** [review]
Approved, as there are only small suggestions left!

**Review by e1Ru1o — APPROVED** [review]

**Pull request merged**